File: android/app/src/main/java/com/airmate/android/ui/mont/MontControls.kt
package com.airmate.android.ui.mont

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Alignment as UiAlignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * The wordmark: the lightest weight over the heaviest at one size. That contrast is the logo.
 */
@Composable
fun MontWordmark(size: Int = 38) {
    val scaled = (size * LocalMontScale.current).sp
    Row {
        Text("air", color = Color.White, fontFamily = Mont, fontWeight = FontWeight.Thin, fontSize = scaled)
        Text("Mate", color = Color.White, fontFamily = Mont, fontWeight = FontWeight.Black, fontSize = scaled)
    }
}

/**
 * A row states its own state through brightness alone. No chevron, no highlight fill: the active
 * item is simply the bright one.
 */
@Composable
fun MontRow(
    label: String,
    modifier: Modifier = Modifier,
    trailing: String? = null,
    enabled: Boolean = true,
    active: Boolean = true,
    onClick: () -> Unit
) {
    Row(
        modifier
            .fillMaxWidth()
            .clickable(enabled = enabled, onClick = onClick)
            .padding(vertical = 7.dp * LocalMontScale.current),
        verticalAlignment = Alignment.CenterVertically
    ) {
        MontLabel(
            label.uppercase(),
            Modifier.weight(1f),
            alpha = when {
                !enabled -> MontWhite.DISABLED
                active -> MontWhite.ACTIVE
                else -> MontWhite.DIM
            }
        )
        trailing?.let { MontLabel(it.uppercase(), alpha = MontWhite.DIM, size = 11) }
    }
}

@Composable
fun MontLabel(
    text: String,
    modifier: Modifier = Modifier,
    alpha: Float = MontWhite.ACTIVE,
    size: Int = 15
) {
    Text(
        text,
        modifier = modifier,
        color = Color.White.copy(alpha),
        fontFamily = Mont,
        fontWeight = FontWeight.Black,
        fontSize = (size * LocalMontScale.current).sp,
        maxLines = 1
    )
}

/** An explanatory line under a row. */
@Composable
fun MontDetail(text: String, modifier: Modifier = Modifier, alpha: Float = MontWhite.DETAIL) {
    Text(
        text,
        modifier = modifier,
        color = Color.White.copy(alpha),
        fontFamily = Mont,
        fontWeight = FontWeight.Normal,
        fontSize = (11 * LocalMontScale.current).sp,
        lineHeight = (15 * LocalMontScale.current).sp
    )
}

/**
 * A choice. No pill, no border, no fill — selected is bright, unselected is dim, exactly the rule
 * a row follows, because a choice *is* a row that happens to sit beside others.
 */
@Composable
fun MontChips(
    options: List<String>,
    selected: Int,
    modifier: Modifier = Modifier,
    onSelect: (Int) -> Unit
) {
    val scale = LocalMontScale.current
    Row(modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(14.dp * scale)) {
        options.forEachIndexed { index, option ->
            Text(
                option.uppercase(),
                modifier = Modifier
                    .clickable { onSelect(index) }
                    .padding(vertical = 4.dp * scale),
                color = Color.White.copy(if (index == selected) MontWhite.ACTIVE else MontWhite.DIM),
                fontFamily = Mont,
                fontWeight = FontWeight.Black,
                fontSize = (11 * scale).sp
            )
        }
    }
}

/**
 * The slider stopped at two positions: a white block filling one half, with the state written in
 * the half it has left.
 *
 * The word names what the control currently is, not what tapping it would do — a switch labelled
 * with its own opposite is a puzzle every single time you meet it.
 */
@Composable
fun MontToggle(
    checked: Boolean,
    onChange: (Boolean) -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true
) {
    val slide by animateFloatAsState(if (checked) 0f else 1f, label = "montToggle")
    val alpha = if (enabled) 1f else MontWhite.DISABLED

    val scale = LocalMontScale.current
    Box(
        modifier
            .width(56.dp * scale)
            .height(18.dp * scale)
            .clickable(enabled = enabled) { onChange(!checked) }
    ) {
        Canvas(Modifier.fillMaxSize()) {
            val half = size.width * .5f
            drawRect(Color.White.copy(MontWhite.TRACK * alpha), Offset.Zero, size)
            drawRect(Color.White.copy(alpha), Offset(half * slide, 0f), Size(half, size.height))
        }
        Text(
            if (checked) "ON" else "OFF",
            modifier = Modifier
                .align(if (checked) Alignment.CenterEnd else Alignment.CenterStart)
                .width(28.dp * scale),
            color = Color.White.copy(alpha),
            fontFamily = Mont,
            fontWeight = FontWeight.Black,
            fontSize = (10 * scale).sp,
            textAlign = TextAlign.Center,
            maxLines = 1
        )
    }
}

/**
 * A Mont card.
 *
 * Square, black at 92%, no border and no shadow — the black is the entire separation. Text hangs
 * off a generous left margin and nothing needs the right one, which is why the padding is
 * asymmetric.
 */
@Composable
fun MontCard(
    modifier: Modifier = Modifier,
    content: @Composable ColumnScope.() -> Unit
) {
    val scale = LocalMontScale.current
    Column(
        modifier
            .background(MontSurface)
            .verticalScroll(rememberScrollState())
            .padding(
                start = 22.dp * scale,
                top = 20.dp * scale,
                end = 14.dp * scale,
                bottom = 16.dp * scale
            ),
        verticalArrangement = Arrangement.spacedBy(1.dp * scale),
        content = content
    )
}

/**
 * Where a Mont card stands on a tablet.
 *
 * One place decides how big the card is and how big the language inside it is, so the pairing
 * card, the onboarding card and the control card cannot disagree. The card is about a quarter of
 * the screen tall and is never allowed past six sevenths of it, scrolling inside that bound —
 * a card whose content decides whether it still fits on the display is a card that eventually
 * does not.
 */
@Composable
fun MontStage(
    alignment: UiAlignment,
    modifier: Modifier = Modifier,
    content: @Composable androidx.compose.foundation.layout.ColumnScope.() -> Unit
) {
    BoxWithConstraints(modifier.fillMaxSize()) {
        // 560dp is roughly the short edge of the phone the language was drawn on, so a tablet
        // lands somewhere near 1.5x and a small screen stays exactly as Mont specifies.
        val scale = (maxHeight / 560.dp).coerceIn(1f, 2f)
        val cardWidth = (maxWidth * 0.44f).coerceIn(340.dp, 760.dp)
        CompositionLocalProvider(LocalMontScale provides scale) {
            MontCard(
                Modifier
                    .align(alignment)
                    .width(cardWidth)
                    .heightIn(min = maxHeight * 0.25f, max = maxHeight * 0.86f),
                content = content
            )
        }
    }
}

File: windows/AirMateWindows/MontControls.cs
namespace AirMate;

/// <summary>
/// A row of the window: a word, full width, that does something.
/// </summary>
/// <remarks>
/// Bright when it can be used and dim when it cannot. There is no box, pill or border and there is
/// not meant to be one — under Mont the type is what makes a word act as a button, and putting a
/// container back is the single thing the language exists to do without.
/// </remarks>
internal sealed class MontRow : Control
{
    private bool hot;

    public string Trailing { get; set; } = string.Empty;
    public bool Dim { get; set; }

    public MontRow(string text, float size = 13f)
    {
        Text = text;
        Font = MontFont.Black(size);
        ForeColor = Mont.Active;
        BackColor = Color.Transparent;
        Height = (int)(size * 2.6f);
        Cursor = Cursors.Hand;
        SetStyle(ControlStyles.AllPaintingInWmPaint | ControlStyles.OptimizedDoubleBuffer
            | ControlStyles.UserPaint | ControlStyles.SupportsTransparentBackColor, true);
    }

    protected override void OnMouseEnter(EventArgs e) { hot = true; Invalidate(); base.OnMouseEnter(e); }
    protected override void OnMouseLeave(EventArgs e) { hot = false; Invalidate(); base.OnMouseLeave(e); }
    protected override void OnEnabledChanged(EventArgs e) { Invalidate(); base.OnEnabledChanged(e); }

    protected override void OnPaint(PaintEventArgs e)
    {
        var colour = !Enabled ? Mont.Disabled : Dim && !hot ? Mont.Dim : Mont.Active;
        TextRenderer.DrawText(e.Graphics, Text.ToUpperInvariant(), Font,
            new Rectangle(0, 0, Width, Height), colour,
            TextFormatFlags.Left | TextFormatFlags.VerticalCenter | TextFormatFlags.NoPrefix);

        if (Trailing.Length > 0)
        {
            TextRenderer.DrawText(e.Graphics, Trailing.ToUpperInvariant(), MontFont.Black(11f),
                new Rectangle(0, 0, Width, Height), Mont.Dim,
                TextFormatFlags.Right | TextFormatFlags.VerticalCenter | TextFormatFlags.NoPrefix);
        }
    }
}

/// <summary>An explanatory line under a row.</summary>
internal sealed class MontDetail : Label
{
    public MontDetail(string text = "", float size = 11f)
    {
        Text = text;
        Font = MontFont.Black(size);
        ForeColor = Mont.Detail;
        AutoSize = false;
        BackColor = Color.Transparent;
    }
}

/// <summary>
/// A choice, as a row of words. Selected is bright, unselected is dim — exactly the rule a row
/// follows, because a choice is a row that happens to sit beside others.
/// </summary>
internal sealed class MontChips : Control
{
    private string[] options = [];
    private int selected = -1;
    private readonly List<Rectangle> hits = [];

    public event Action<int>? SelectionChanged;

    public MontChips()
    {
        BackColor = Color.Transparent;
        Font = MontFont.Black(11f);
        Height = 26;
        Cursor = Cursors.Hand;
        SetStyle(ControlStyles.AllPaintingInWmPaint | ControlStyles.OptimizedDoubleBuffer
            | ControlStyles.UserPaint | ControlStyles.SupportsTransparentBackColor, true);
    }

    public void SetOptions(IEnumerable<string> values, int selectedIndex)
    {
        options = values.Select(v => v.ToUpperInvariant()).ToArray();
        selected = selectedIndex;
        Invalidate();
    }

    public int Selected => selected;

    protected override void OnPaint(PaintEventArgs e)
    {
        hits.Clear();
        int x = 0;
        for (int index = 0; index < options.Length; index++)
        {
            var size = TextRenderer.MeasureText(e.Graphics, options[index], Font);
            var bounds = new Rectangle(x, 0, size.Width, Height);
            hits.Add(bounds);
            TextRenderer.DrawText(e.Graphics, options[index], Font, bounds,
                index == selected ? Mont.Active : Mont.Dim,
                TextFormatFlags.Left | TextFormatFlags.VerticalCenter | TextFormatFlags.NoPrefix);
            x += size.Width + 16;
        }
    }

    protected override void OnMouseDown(MouseEventArgs e)
    {
        for (int index = 0; index < hits.Count; index++)
        {
            if (!hits[index].Contains(e.Location)) continue;
            selected = index;
            Invalidate();
            SelectionChanged?.Invoke(index);
            break;
        }
        base.OnMouseDown(e);
    }
}

/// <summary>
/// The slider stopped at two positions: a white block filling one half, with the state written in
/// the half it has left.
/// </summary>
/// <remarks>
/// The word names what the control currently is, not what pressing it would do. A switch labelled
/// with its own opposite is a puzzle every single time you meet it. Windows' own CheckBox is a
/// rounded, bordered, animated object — every decoration Mont removed — so it is not used here.
/// </remarks>
internal sealed class MontToggle : Control
{
    private bool isOn;

    public event Action<bool>? Toggled;

    public MontToggle()
    {
        Size = new Size(56, 18);
        BackColor = Color.Transparent;
        Font = MontFont.Black(10f);
        Cursor = Cursors.Hand;
        SetStyle(ControlStyles.AllPaintingInWmPaint | ControlStyles.OptimizedDoubleBuffer
            | ControlStyles.UserPaint | ControlStyles.SupportsTransparentBackColor, true);
    }

    public bool IsOn
    {
        get => isOn;
        set { isOn = value; Invalidate(); }
    }

    protected override void OnPaint(PaintEventArgs e)
    {
        int half = Width / 2;
        using var track = new SolidBrush(Enabled ? Mont.Track : Color.FromArgb(12, 255, 255, 255));
        using var block = new SolidBrush(Enabled ? Mont.Active : Mont.Disabled);
        e.Graphics.FillRectangle(track, 0, 0, Width, Height);
        e.Graphics.FillRectangle(block, isOn ? 0 : half, 0, half, Height);

        var text = isOn ? "ON" : "OFF";
        var bounds = isOn ? new Rectangle(half, 0, half, Height) : new Rectangle(0, 0, half, Height);
        TextRenderer.DrawText(e.Graphics, text, Font, bounds,
            Enabled ? Mont.Active : Mont.Disabled,
            TextFormatFlags.HorizontalCenter | TextFormatFlags.VerticalCenter | TextFormatFlags.NoPrefix);
    }

    protected override void OnMouseDown(MouseEventArgs e)
    {
        if (Enabled)
        {
            isOn = !isOn;
            Invalidate();
            Toggled?.Invoke(isOn);
        }
        base.OnMouseDown(e);
    }
}

/// <summary>A measurement: a small dim label over the figure it names.</summary>
internal sealed class MontMetric : Control
{
    public string Title { get; set; } = string.Empty;
    public string Value { get; set; } = string.Empty;

    public MontMetric(string title)
    {
        Title = title;
        BackColor = Color.Transparent;
        Size = new Size(120, 40);
        SetStyle(ControlStyles.AllPaintingInWmPaint | ControlStyles.OptimizedDoubleBuffer
            | ControlStyles.UserPaint | ControlStyles.SupportsTransparentBackColor, true);
    }

    protected override void OnPaint(PaintEventArgs e)
    {
        TextRenderer.DrawText(e.Graphics, Title.ToUpperInvariant(), MontFont.Black(9f),
            new Rectangle(0, 0, Width, 14), Mont.Dim, TextFormatFlags.Left | TextFormatFlags.NoPrefix);
        TextRenderer.DrawText(e.Graphics, Value, MontFont.Black(13f),
            new Rectangle(0, 14, Width, 22), Mont.Primary, TextFormatFlags.Left | TextFormatFlags.NoPrefix);
    }
}

/// <summary>The wordmark: the lightest weight over the heaviest at one size.</summary>
internal sealed class MontWordmark : Control
{
    private readonly float size;

    public MontWordmark(float size = 28f)
    {
        this.size = size;
        BackColor = Color.Transparent;
        Height = (int)(size * 1.6f);
        SetStyle(ControlStyles.AllPaintingInWmPaint | ControlStyles.OptimizedDoubleBuffer
            | ControlStyles.UserPaint | ControlStyles.SupportsTransparentBackColor, true);
    }

    protected override void OnPaint(PaintEventArgs e)
    {
        using var thin = MontFont.Thin(size);
        using var black = MontFont.Black(size);
        var airWidth = TextRenderer.MeasureText(e.Graphics, "air", thin,
            new Size(int.MaxValue, int.MaxValue), TextFormatFlags.NoPadding).Width;
        TextRenderer.DrawText(e.Graphics, "air", thin, new Point(0, 0), Mont.Active, TextFormatFlags.NoPadding);
        TextRenderer.DrawText(e.Graphics, "Mate", black, new Point(airWidth, 0), Mont.Active, TextFormatFlags.NoPadding);
    }
}
